Rest API Learning

gokulhk opened the pull request:
Rest API Learning practise code:

1. CRUD API endpoints for bookmarks and folders.

gokulhk committed:
chore: default config files

gokulhk committed:
feat: added rest api learning folder

gokulhk committed:
feat: added PUT & PATCH functionality

gokulhk committed:
feat: added api versioning

gokulhk committed:
feat: added folder support for bookmarks

gokulhk committed:
Remove .idea files from tracking

gokulhk committed:
feat: added folder name support

mbnsky commented on rest-api/src/main/java/com/example/rest_api/RestApiApplication.java:
```diff
@@ -0,0 +1,12 @@
+package com.example.rest_api;
```
Please align package naming to https://google.github.io/styleguide/javaguide.html#s5.2.1-package-names
which is the most commonly used approach

gokulhk replied:
updated it, thanks for letting me know.

mbnsky commented on rest-api/src/main/java/com/example/rest_api/controllers/BookmarkController.java:
```diff
@@ -0,0 +1,100 @@
+  }
+
+  @PostMapping(
+      consumes = MediaType.APPLICATION_JSON_VALUE,
```
Why did you add `consumes` and `produces`?

gokulhk replied:
By default it allows all formats to map to request body so to make the api contract clear I added these to specify what is expected and what is returned. Please let me know your thoughts.

mbnsky replied:
If you really want to restrict it this way you can. No reason for overcomplicating for now if not needed.
Additionally as far as I remember if you set `consumes = MediaType.APPLICATION_JSON_VALUE` then it won't match content type like `application/json; charset=utf-8`

gokulhk replied:
ok

mbnsky commented on rest-api/src/main/java/com/example/rest_api/controllers/BookmarkController.java:
```diff
@@ -0,0 +1,100 @@
+@Log
+public class BookmarkController {
+
+  @Autowired BookmarkRepository bookmarkRepository;
```
Please make use of constructor based injection instead of field based injection

gokulhk replied:
Updated it.

mbnsky commented on rest-api/src/main/java/com/example/rest_api/controllers/FolderController.java:
```diff
@@ -0,0 +1,63 @@
+@RestController
+@RequestMapping("/api/v1/folders")
+@Log
+public class FolderController {
```
It's not in the scope of tasks 1&2, right? 😄

gokulhk replied:
yeah @mbnsky . I thought to keep up with weekly tasks so added it in the same pr. I've removed it now.

mbnsky commented on rest-api/src/main/java/com/example/rest_api/entities/Folder.java:
```diff
@@ -0,0 +1,30 @@
+  private String id;
+
+  private String name;
+  private List<Bookmark> bookmarkList;
```
why `bookmarkList` instead of just `bookmarks`?

gokulhk replied:
yes, it has to `bookmarks`. My bad 😅, I missed to update it before submitting for review.

mbnsky reviewed (commented):
Please add tests and cover only tasks 1&2 with this PR.
I'll take a look at it again after that

gokulhk committed:
revert: folder feature & review changes

gokulhk committed:
test: added unit tests

gokulhk commented:
> Please add tests and cover only tasks 1&2 with this PR. I'll take a look at it again after that

Hey @mbnsky , I've added unit tests and updated code as suggested. Please take a look at the changes when possible.

mbnsky commented on rest-api/src/test/java/com/example/restapi/controllers/BookmarkControllerTest.java:
```diff
@@ -0,0 +1,132 @@
+            MockMvcRequestBuilders.put("/api/v1/bookmarks/1")
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(
+                    "{\"id\":\"1\",\"title\":\"UpdatedTitle\",\"description\":\"UpdatedDescription\",\"url\":\"http://example.com/updated\"}"))
```
json would be way more readable if we would use text block here like
```
String content = """
{
  "someProperty": "someValue"
}
""";
```

gokulhk replied:
sure, I'll update it.

gokulhk replied:
updated it.

mbnsky commented on rest-api/src/main/java/com/example/restapi/controllers/BookmarkController.java:
```diff
@@ -0,0 +1,101 @@
+
+    if (bookmarkOptional.isEmpty()) return ResponseEntity.notFound().build();
+
+    Bookmark bookmark = bookmarkOptional.get();
```
I'll accept it for now but I'll ask you to move that kind of logic to another layer like `BookmarksService`.

gokulhk replied:
Yeah, agreed @mbnsky .

gokulhk replied:
Hey @mbnsky , I've moved service logic to service layer and updated unit tests accordingly.

gokulhk committed:
refactor: updated test payload construction logic

gokulhk committed:
build: moved service logic to service layer

gokulhk committed:
feat: added global exception handling

gokulhk committed:
tests: added controller & service layer tests
